### PR TITLE
FP-2468 - Study performance decrease in tree view flow editor

### PR DESCRIPTION
**Pull request by quirinpa**
- [FP-2468 - Study performance decrease in tree view flow editor](https://movai.atlassian.net/browse/FP-2468)

### CODE REVIEW

**Comment by quirinpa** [comment]
🥳🎊🚀🎉

File: src/editors/Flow/view/Components/interface/canvas.js [hunk 1]
```diff
@@ -704,10 +692,11 @@ class Canvas {
       .duration(750)
       .call(
         this.zoomBehavior.transform,
-        d3.zoomIdentity
+        (d3.zoomIdentity
           .translate(width / 2, height / 2)
-          .translate(-SCALE * xCoordinate, -SCALE * yCoordinate)
+          .translate(- SCALE * xCoordinate, - SCALE * yCoordinate)
```
Comment (manuelnogueiramov) [inline src/editors/Flow/view/Components/interface/canvas.js:697]: I'm curious, doesn't this space change the calculation?

File: src/editors/Flow/view/Flow.jsx [hunk 2]
```diff
@@ -89,14 +104,12 @@ export const Flow = (props, ref) => {
...
   // State Hooks
-  const [loading, setLoading] = useState(true);
   const [dataFromDB, setDataFromDB] = useState();
   const [robotSelected, setRobotSelected] = useState("");
   const [runningFlow, setRunningFlow] = useState("");
   const [warnings, setWarnings] = useState([]);
   const [flowDebugging, setFlowDebugging] = useState();
-  const [warningsVisibility, setWarningsVisibility] = useState(false);
-  const [viewMode, setViewMode] = useState(FLOW_VIEW_MODE.default);
+  const viewMode = mainInterface?.viewMode ?? "default";
```
Comment (manuelnogueiramov) [inline src/editors/Flow/view/Flow.jsx:112]: This is no longer a state hook, so, on every render this `viewMode` variable will be recreated.

File: src/plugins/DocManager/DocManager.js [hunk 3]
```diff
@@ -141,7 +141,15 @@ class DocManager extends IDEPlugin {
...
-    return this.getStore(scope)?.discardDocChanges(name);
+
+    const ret = this.getStore(scope)?.discardDocChanges(name)
+
+    this.emit(PLUGINS.DOC_MANAGER.ON.DISCARD_DOC, {
+      name,
+      scope,
+    });
+
+    return ret;
```
Comment (manuelnogueiramov) [inline src/plugins/DocManager/DocManager.js:152]: Any reason why the discard needs to be done before the emit?

```suggestion

    this.emit(PLUGINS.DOC_MANAGER.ON.DISCARD_DOC, {
      name,
      scope,
    });

    return this.getStore(scope)?.discardDocChanges(name);
```

**Review by manuelnogueiramov — CHANGES_REQUESTED** [review]
To be fair I saw a couple of improvements, ie: when selecting node and then selecting another node, there's no longer a 300ms delay (which is very good, even though it wasn't noticeable by most humans). This, just to say that the idea of the PR seems good and the direction seems interesting. 
The issue is, I made a couple of comments, to be honest most of them aren't that big of a deal. But upon testing the PR thoroughly I found a lot of regressions that need to be addressed before merging this. 

List of regressions that I've found (which doesn't mean more might have slipped):
- No warnings being shown.
no_warnings_shown

- After dragging a selection of nodes, or a selected node it gets de-selected in the end.
shift_drag_unselects

- Double clicking a subflow, doesn't open said subflow (infinite loading).
open_subflow_explodes

- Double clicking the start node in tree view, causes Add Node Bookmark to forever disappear.double_click_start_node

- Selecting a node, or multiple nodes, causes it/them to be dragged along side with another node that you want to drag.
drag_weird_selection

- Update document is no longer working.
not_updating_flow


As I said, there might be a lot more regression issues I didn't catch.

File: src/editors/Flow/view/Components/interface/canvas.js (hunk 1 above)
Comment (quirinpa) [reply src/editors/Flow/view/Components/interface/canvas.js:697]: I don't think so.

File: src/editors/Flow/view/Flow.jsx (hunk 2 above)
Comment (quirinpa) [reply src/editors/Flow/view/Flow.jsx:112]: Ok, we can use useMemo to avoid that.

File: src/plugins/DocManager/DocManager.js (hunk 3 above)
Comment (quirinpa) [reply src/plugins/DocManager/DocManager.js:152]: I guess this has to be tried out.

**Comment by quirinpa** [comment]
>     * Update document is no longer working.
>       not_updating_flow

Just to clarify, it is not the "saving" that is in question here. The document is saved.
The following video shows how in version 2.4.0-61 of project tugbot there was already a problem with "Update document".
Suggesting that this issue is not fully introduced by this PR:
untitled

**Comment by quirinpa** [comment]
> * After dragging a selection of nodes, or a selected node it gets de-selected in the end.
This is fixed.

**Comment by quirinpa** [comment]
> * No warnings being shown.
This problem was already in dev. This PR fixes it:
https://github.com/MOV-AI/frontend-npm-lib-ide/pull/173

File: src/plugins/DocManager/DocManager.js (hunk 3 above)
Comment (quirinpa) [reply src/plugins/DocManager/DocManager.js:152]: Yes. There is a reason for it.

**Comment by quirinpa** [comment]
> * Double clicking the start node in tree view, causes Add Node Bookmark to forever disappear.

Not anymore ;)

**Comment by quirinpa** [comment]
> * Selecting a node, or multiple nodes, causes it/them to be dragged along side with another node that you want to drag.

Fixed

**Comment by quirinpa** [comment]
> > > ```
> > > * Double clicking a subflow, doesn't open said subflow (infinite loading).
> > >   open_subflow_explodes
> > > ```
> > 
> > 
> > This is not happening to me. It may be a (older) problem that happens sometimes in which the flow fails to load due to a timeout. Can you please test again and see if it still happens?
> 
> I could replicate this 100%. As in, if I open the flow from the Explorer, it opens fine, but when I try to open it from inside the flow it will happen 100%. Also, tested in 2.4.0-4 and it worked fine (to be fair didn't test in dev yet to check if it isn't caused by this PR).

Ok, but:
noInfinite
Certainly not 100% of the time, for me.

**Comment by quirinpa** [comment]
> > > ```
> > > * Update document is no longer working.
> > >   not_updating_flow
> > > ```
> > 
> > 
> > Just to clarify, it is not the "saving" that is in question here. The document is saved. The following video shows how in version 2.4.0-61 of project tugbot there was already a problem with "Update document". Suggesting that this issue is not fully introduced by this PR: untitled
> >       
> >         [
> >           
> >         
> >             untitled
> >           ](https://user-images.githubusercontent.com/4276265/255618133-e3a51484-7a4a-4d44-b445-e11f6c938658.gif)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://user-images.githubusercontent.com/4276265/255618133-e3a51484-7a4a-4d44-b445-e11f6c938658.gif)
> 
> You're right. I can confirm that, and already have a fix [here](https://github.com/MOV-AI/frontend-npm-lib-ide/pull/174).
I didn't realize you were trying to fix that in that PR. I thought that was for Michal's bugs.

**Pull request closed without merging**